pkg/account/deployer/client.go

```diff
@@ -0,0 +1,268 @@
+			return "", err
+		}
+		return createdPolicy.GetUuid(), nil
+	}
+
+	if len(existingPolicies) > 1 { // shouldn't happen
+		log.Warn("Found multiple policies with name %q. Updating policy with UUID %q", policy.Name, existingPolicies[0].GetUuid())
+	}
+
+	log.Debug("Trying to update existing policy with name %q and UUID %q", policy.Name, existingPolicies[0].GetUuid())
+	_, resp, err = d.client.PolicyManagementAPI.UpdateLevelPolicy(ctx, policyLevel, policyLevelId, existingPolicies[0].GetUuid()).CreateOrUpdateLevelPolicyRequestDto(policy).Execute()
+	if err = d.handleClientResponseError(resp, err, "unable to update policy with name: "+policy.Name); err != nil {
+		return "", err
+	}
+	return existingPolicies[0].GetUuid(), nil
+}
+
+func (d *accountManagementClient) upsertGroup(ctx context.Context, groupId string, group Group) (remoteId, error) {
+	if groupId != "" {
+		log.Debug("Trying to update policy with origin object ID (UUID) %q", groupId)
+		resp, err := d.client.GroupManagementAPI.EditGroup(ctx, d.accountInfo.AccountUUID, groupId).PutGroupDto(group).Execute()
+
+		if err = d.handleClientResponseError(resp, err, "unable to update policy with UUID: "+groupId); err != nil {
+			return "", err
+		}
+		return groupId, nil
+	}
+
+	result, resp, err := d.client.GroupManagementAPI.GetGroups(ctx, d.accountInfo.AccountUUID).Execute()
+	if err = d.handleClientResponseError(resp, err, "unable to get groups with name: "+group.Name); err != nil {
+		return "", err
+	}
+
+	// find groups with matching name
+	var existingGroups []accountmanagement.GetGroupDto
+	for _, g := range result.GetItems() {
+		if g.GetName() == group.Name {
+			existingGroups = append(existingGroups, g)
+		}
+	}
+
+	if len(existingGroups) == 0 {
+		var createdGroups []accountmanagement.GetGroupDto
+		createdGroups, resp, err = d.client.GroupManagementAPI.CreateGroups(ctx, d.accountInfo.AccountUUID).PutGroupDto([]accountmanagement.PutGroupDto{group}).Execute()
+		if err = d.handleClientResponseError(resp, err, "unable to create group with name: "+group.Name); err != nil {
+			return "", err
+		}
+		if len(createdGroups) < 1 {
+			return "", fmt.Errorf("unable to get UUID of created group with name: %s", group.Name)
+		}
+		return createdGroups[0].GetUuid(), nil
+	}
+
+	if len(existingGroups) > 1 {
+		log.Warn("Updating multiple policies with name %s", group.Name)
+	}
+
+	groupToUpdate := existingGroups[0]
+
+	resp, err = d.client.GroupManagementAPI.EditGroup(ctx, d.accountInfo.AccountUUID, groupToUpdate.GetUuid()).PutGroupDto(group).Execute()
+	if err = d.handleClientResponseError(resp, err, "unable to update group with name: "+group.Name); err != nil {
+		return "", err
+	}
+	return groupToUpdate.GetUuid(), nil
+}
+
+func (d *accountManagementClient) upsertUser(ctx context.Context, userId string) (remoteId, error) {
+	users, resp, err := d.client.UserManagementAPI.GetUsers(ctx, d.accountInfo.AccountUUID).Execute()
```

Comment:
How does the API behave if we just try to create a user that already exist? I guess we can't differentiate an actual error from "already exist" and thus need to do the GET first?

As the email is the ID we could just query for the exact user in that case though, instead of getting all and iterating :)

Reply:
it was done like that under the assumption that the user ID is a UUDI or so. :D 
but it's the email address, so you are right

Reply:
but on the other hand. the api returns a 400 BadRequest err code (instead of 409 Conflict) so I cannot really count on that.. also there is no possibility for querying exactly one user. :/
so i guess i will keep it like it is...

Reply:
There is [/iam/v1/accounts/{accountUuid}/users/{email}](https://api.dynatrace.com/spec/#/User%20management/UsersController_getUserGroups) - which is described as "list groups of a user" but behaves as the resource path suggests and returns the user object or 404 if the user does not exist.

We could call that and simply check if it returned a success code or 404